### PR TITLE
Users

### PR DESCRIPTION
**Pull request by osichan**

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -0,0 +1,6 @@
+SECRET_KEY='d1d$y!^0m%uuxjhbuzlycl3a)veoc(mu$_v1!qkx^^)_o5f&no'
+DEBUG='1'
+DB_NAME='fake_instagram'
+DB_USER='postgres'
+DB_PASSWORD='Irsik2005'
+REDIS_PASSWORD='randomPassword'
```
Comment (MarkoYaminsky) [inline .env]: gitignore your `.env`

File: backend/.gitignore [hunk 2]
```diff
@@ -0,0 +1,4 @@
+venv/
+server/static/
+**/__pycache__/
+server/erp_backend/companies.json
```
Comment (MarkoYaminsky) [inline backend/.gitignore]: why do you have two gitignores, if you are ony writing backend in this project?

File: backend/instagram_backend/apps/users/models.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    def create_user(self, email, name, username, password, bio='', profile_picture=None):
+        user = self.model(
+            email=email,
+            name=name,
+            username=username,
+            password=password,
+            bio=bio,
+            profile_picture=profile_picture
+        )
+        user.save(using=self._db)
```
Comment (MarkoYaminsky) [inline backend/instagram_backend/apps/users/models.py]: unless you are using multiple databases, there is no need to specify which one you are using

File: backend/instagram_backend/apps/users/models.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
+from django.db import models
+
+
+class UserManager(models.Manager):
+    def create_user(self, email, name, username, password, bio='', profile_picture=None):
```
Comment (MarkoYaminsky) [inline backend/instagram_backend/apps/users/models.py]: add type hinting

File: backend/instagram_backend/apps/users/models.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+            name=name,
+            username=username,
+            password=password,
+            bio=bio,
+            profile_picture=profile_picture
+        )
+        user.save(using=self._db)
+        return user
+
+    def update_user(self, user_id, **kwargs):
```
Comment (MarkoYaminsky) [inline backend/instagram_backend/apps/users/models.py]: add type hinting

File: backend/instagram_backend/apps/users/models.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        user = self.get(pk=user_id)
+        for attr, value in kwargs.items():
+            setattr(user, attr, value)
+        user.save(using=self._db)
+        return user
+
+    def delete_user(self, user_id):
```
Comment (MarkoYaminsky) [inline backend/instagram_backend/apps/users/models.py]: add type hinting

File: backend/instagram_backend/apps/users/models.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        user = self.get(pk=user_id)
+        user.delete()
+
+
+class FollowerManager(models.Manager):
+    def add_follower(self, follower_id, followee_id):
```
Comment (MarkoYaminsky) [inline backend/instagram_backend/apps/users/models.py]: add type hinting

File: backend/instagram_backend/apps/users/models.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+        follower = self.model(
+            follower_id=follower_id,
+            followee_id=followee_id
+        )
+        follower.save(using=self._db)
+        return follower
+
+    def remove_follower(self, follower_id, followee_id):
```
Comment (MarkoYaminsky) [inline backend/instagram_backend/apps/users/models.py]: add type hinting

File: backend/instagram_backend/apps/users/models.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        try:
+            follower = self.get(follower_id=follower_id, followee_id=followee_id)
+            follower.delete()
+        except self.model.DoesNotExist:
+            pass
+
+
+class User(models.Model):
+    objects = UserManager()
+    email = models.CharField(max_length=255, unique=True, null=False)
```
Comment (MarkoYaminsky) [inline backend/instagram_backend/apps/users/models.py]: `null=False` is an implicit option. Correct this here and in the rest of model fields

File: backend/instagram_backend/apps/users/models.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+    name = models.CharField(max_length=45)
+    username = models.CharField(max_length=45, unique=True, null=False)
+    password = models.CharField(max_length=45, null=False)
+    bio = models.CharField(max_length=255)
```
Comment (MarkoYaminsky) [inline backend/instagram_backend/apps/users/models.py]: If I am not mistaken, the default value of `bio` was `""` in your `create_user` method. So if you want to store blank string, you should give option `blank=True` to this `CharField`

File: backend/instagram_backend/apps/users/models.py [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    profile_picture = models.IntegerField()
```
Comment (MarkoYaminsky) [inline backend/instagram_backend/apps/users/models.py]: Why is this `IntegerField`?

File: backend/instagram_backend/apps/users/models.py [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+
+    def __str__(self):
+        return self.username
+
+
+class Follower(models.Model):
+    objects = FollowerManager()
+    follower_id = models.ForeignKey(User, on_delete=models.CASCADE, related_name='followers')
```
Comment (MarkoYaminsky) [inline backend/instagram_backend/apps/users/models.py]: you don't name fields with `id` in the end. Cause follower would be an object, and `follower_id` is just this object's id. Correct this here and below

File: backend/instagram_backend/apps/users/views.py [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+from models import User, Follower
+from serializers import UserSerializer, FollowerSerializer
+
+
+class UserListView(generics.ListAPIView):
+    queryset = User.objects.all()
+    serializer_class = UserSerializer
+
+    def get_queryset(self):
+        return User.objects.values('username', 'profile_picture')
```
Comment (MarkoYaminsky) [inline backend/instagram_backend/apps/users/views.py]: why are you getting values out of this queryset? If you want only `username` and `profile_picture` in your response, you should do this on serializer level, not here

File: backend/instagram_backend/instagram_backend/asgi.py [hunk 14]
```diff
@@ -0,0 +1,16 @@
+"""
+ASGI config for instagram_backend project.
+
+It exposes the ASGI callable as a module-level variable named ``application``.
+
+For more information on this file, see
+https://docs.djangoproject.com/en/4.2/howto/deployment/asgi/
+"""
```
Comment (MarkoYaminsky) [inline backend/instagram_backend/instagram_backend/asgi.py]: delete these strings (here and below)

File: backend/instagram_backend/instagram_backend/settings.py [hunk 15]
```diff
@@ -0,0 +1,114 @@
+import os
+from pathlib import Path
+from dotenv import load_dotenv
+
+load_dotenv()
+
+# Build paths inside the project like this: BASE_DIR / 'subdir'.
```
Comment (MarkoYaminsky) [inline backend/instagram_backend/instagram_backend/settings.py]: delete these comments (here and below)

File: backend/instagram_backend/instagram_backend/urls.py [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+    1. Add an import:  from other_app.views import Home
+    2. Add a URL to urlpatterns:  path('', Home.as_view(), name='home')
+Including another URLconf
+    1. Import the include() function: from django.urls import include, path
+    2. Add a URL to urlpatterns:  path('blog/', include('blog.urls'))
+"""
+from django.contrib import admin
+from django.urls import path
+
+urlpatterns = [
```
Comment (MarkoYaminsky) [inline backend/instagram_backend/instagram_backend/urls.py:6]: did you register your endpoints?

**Review by MarkoYaminsky — COMMENTED** [review]
Overall I would also like to point out that it is strongly recommended to split business logic and data access into two different layers. By that I mean that services like creating, updating, deleting and retrieving, shouldn't be on the same layer that models

File: .env (hunk 1 above)
Comment (smykandriy) [reply .env]: not resolved yet

File: backend/.gitignore (hunk 2 above)
Comment (smykandriy) [reply backend/.gitignore]: server/erp_backend/companies.json is outdated from other project

File: backend/instagram_backend/apps/users/models.py [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+
+
+class UserManager(models.Manager):
+    def create_user(self, email: str, name: str, username: str, password: str, bio: str = '',
+                    profile_picture: int = None):
+        user = self.model(
+            email=email,
+            name=name,
+            username=username,
+            password=password,
```
Comment (smykandriy) [inline backend/instagram_backend/apps/users/models.py]: user.set_password(password)

use this instead

File: backend/instagram_backend/apps/users/models.py [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+            bio=bio,
+            profile_picture=profile_picture
+        )
+        user.save()
+        return user
+
+    def update_user(self, user_id: int, **kwargs):
+        user = self.get(pk=user_id)
+        for attr, value in kwargs.items():
```
Comment (smykandriy) [inline backend/instagram_backend/apps/users/models.py:20]: set_password() to update password

File: backend/instagram_backend/apps/users/models.py [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+        )
+        follower.save(using=self._db)
+        return follower
+
+    def remove_follower(self, followee_id: int, follower_id: int):
+        try:
+            follower = self.get(follower_id=follower_id, followee_id=followee_id)
+            follower.delete()
+        except self.model.DoesNotExist:
+            pass
```
Comment (smykandriy) [inline backend/instagram_backend/apps/users/models.py:46]: better to add some description for exception

**Review by smykandriy — CHANGES_REQUESTED** [review]

**Review by smykandriy — APPROVED** [review]

**Pull request merged**